### PR TITLE
Fix/history wrong variants

### PR DESCRIPTION
**Pull request by marikaris**
Some variants were incorrectly presented:
[https://github.com/molgenis/data-transform-vkgl/issues/14](url)
By fixing this issue, the history of those variants is lost. To fix this as much as possible, we try and find variants with the same gene,c_dna, and transcript to connect the history to the actual variant. For variants that don't have this information, we accept that history is lost. The variants were incorrect and don't make any sense, so we want to keep them out of our database. 
Variants with incorrect names found in previous exports, will be written to a csv file with a nice message. This file can be used to write a script that puts a note in the comments of the invalid variants in the history table telling that the variant is invalid and the id of the correct variant in future releases.

### CODE REVIEW

File: consensus/ConsensusFileGenerator.py [hunk 1]
```diff
@@ -21,9 +21,14 @@ def __init__(self, data, tables):
         comments_table = tables['comments_table']
         self.consensus_data = data['consensus_data']
         self.lab_classifications = data['lab_classifications']
-        self.history = data['history']
+        self.history = data['history']['history']
```
Comment (dennishendriksen) [inline consensus/ConsensusFileGenerator.py:24]: what is the history of the history?

File: consensus/ConsensusFileGenerator.py [hunk 2]
```diff
@@ -21,9 +21,14 @@ def __init__(self, data, tables):
...
+        self.alternative_history = data['history']['alternative']
```
Comment (dennishendriksen) [inline consensus/ConsensusFileGenerator.py:25]: what is an alternative history?

File: consensus/ConsensusFileGenerator.py [hunk 3]
```diff
@@ -21,9 +21,14 @@ def __init__(self, data, tables):
...
         self.consensus_table_file_name = consensus_table
         self.comments_table_file_name = comments_table
+        self.incorrect_variant_history_file_name = incorrect_variant_history_file
+        if self.incorrect_variant_history_file_name:
+            incorrect_history_file = open(self.incorrect_variant_history_file_name, 'w')
```
Comment (dennishendriksen) [inline consensus/ConsensusFileGenerator.py:30]: what is the purpose of opening and immediately closing a file?

File: consensus/ConsensusFileGenerator.py [hunk 4]
```diff
@@ -153,7 +158,9 @@ def _get_matching_history(self, variant):
         chromosome = variant["chromosome"]
         gene = variant["gene"]
         variant_type = variant['type']
-
+        incorrect_history_file = None
+        if self.incorrect_variant_history_file_name:
+            incorrect_history_file = open(self.incorrect_variant_history_file_name, 'a')
```
Comment (dennishendriksen) [inline consensus/ConsensusFileGenerator.py:163]: what happens if you already have an existing file (e.g. you execute this piece of code twice)? will you end up with a different file that is twice as large?

File: consensus/ConsensusFileGenerator.py [hunk 5]
```diff
@@ -246,3 +265,9 @@ def generate_consensus_files(self):
         comments_file.close()
         progress_bar.finish()
         return consensus_file_name, comments_file_name
+
+    @staticmethod
+    def _check_alternative_history(transcript, c_dna, gene, export):
```
Comment (dennishendriksen) [inline consensus/ConsensusFileGenerator.py:270]: what is the purpose of `gene`? `transcript:c_dna` should be unique here?

File: consensus/__main__.py [hunk 6]
```diff
@@ -9,12 +11,14 @@
...
 def main():
     # Get data from config
-    config = ConfigParser('config/config.txt') # To run by pressing play in pycharm, use ../config/config.txt
+    config = ConfigParser('config/config.txt')  # To run by pressing play in pycharm, use ../config/config.txt
     consensus_table = config.prefix + config.consensus
     comments_table = config.prefix + config.comments
     molgenis_server = molgenis.Session(config.server)
     history_table = config.history
     previous_exports = config.previous
+    if type(previous_exports) != list:
```
Comment (dennishendriksen) [inline consensus/__main__.py:20]: is this related to the history fix?

File: tests/ConsensusFileGenerator_test.py [hunk 7]
```diff
@@ -46,3 +47,11 @@ def test__get_history_ids_for_variant(self, _, variant_info, expected):
         observed = ConsensusFileGenerator._get_history_ids_for_variant(variant_id, chromosome, pos, ref, alt, gene,
                                                                        variant_type)
         self.assertEqual(observed, expected)
+
+    def test__check_alternative_history(self):
```
Comment (dennishendriksen) [inline tests/ConsensusFileGenerator_test.py:51]: If I understand correctly you have the following cases:
1. before: variant is correct and has transcript:cDNA after: variant is correct
2. before: variant is correct and has no transcript:cDNA after: variant is correct
3. before: variant is incorrect and has transcript:cDNA after: new correct variant
4. before: variant is incorrect and has transcript:cDNA after: merged with other existing correct variant
5. before: variant is incorrect and has no transcript:cDNA after: new correct variant

Are are of them covered by the unit tests?

File: consensus/ConsensusFileGenerator.py [hunk 8]
```diff
@@ -165,6 +172,18 @@ def _get_matching_history(self, variant):
                 variant_history = self._add_history_of_variant(possible_id + '_dup0', export, variant_history)
                 variant_history = self._add_history_of_variant(possible_id + '_dup1', export, variant_history)
 
+            alternative_history = self.alternative_history[export_id]
+            if 'transcript' in variant and 'c_dna' in variant:
+                variant_id = self._check_alternative_history(variant['transcript'], variant['c_dna'], gene,
+                                                             alternative_history)
+                if variant_id and variant_id not in variant_history:
+                    variant_history.append(variant_id)
+                    message = '{} is invalid; will be replaced by correct variant {} in future releases\n'.format(
```
Comment (dennishendriksen) [inline consensus/ConsensusFileGenerator.py:181]: if I run the molgenis-py-consensus pipeline then the incorrect variant will be replaced with a new correct variant (or merged with an existing correct variant) directly, right? so why do you refer to future releases here?

File: consensus/ConsensusFileGenerator.py (hunk 1 above)
Comment (marikaris) [reply consensus/ConsensusFileGenerator.py:24]: This is the history as it was before, the correct. For this fix we had to add an "alternate" history. Variants that have been mapped incorrectly received an incorrect ID. Their cDNA and transcript however are not altered by us, so we can still recognise the variants using these two columns. Those variants are kept in the "alternative history", so these variants have an incorrect id, but can still be recognised using their transcript and cDNA.

File: consensus/ConsensusFileGenerator.py (hunk 2 above)
Comment (marikaris) [reply consensus/ConsensusFileGenerator.py:25]: See comment above

File: consensus/ConsensusFileGenerator.py (hunk 3 above)
Comment (marikaris) [reply consensus/ConsensusFileGenerator.py:30]: By opening a file in the "w" mode, the file will be emptied, this way I make sure the file I append to later, will be empty when I do so.

File: consensus/ConsensusFileGenerator.py (hunk 4 above)
Comment (marikaris) [reply consensus/ConsensusFileGenerator.py:163]: See comment above.

File: consensus/ConsensusFileGenerator.py (hunk 5 above)
Comment (marikaris) [reply consensus/ConsensusFileGenerator.py:270]: On very rare occasions one variant can be assigned to two different genes.

File: consensus/__main__.py (hunk 6 above)
Comment (marikaris) [reply consensus/__main__.py:20]: No, but it is a bug we never found because we never had one history date since we implemented this. I tested this code using only one history date (because the number of dates is not relevant for the fix) and without this line, it will then fail.

File: tests/ConsensusFileGenerator_test.py (hunk 7 above)
Comment (marikaris) [reply tests/ConsensusFileGenerator_test.py:51]: You are correct (except that in case 3 you probably meant "has no transcript:cDNA"). This unit test is only about the alternate history but I agree that I could test for these cases elsewhere. I'll look into that.

File: consensus/ConsensusFileGenerator.py (hunk 8 above)
Comment (marikaris) [reply consensus/ConsensusFileGenerator.py:181]: Yes, the message does look a bit off. I'll change this message.
